runner.go

```diff
@@ -144,6 +144,70 @@ func (d *Runner) Run() error {
 	return err
 }
 
+// Run will validate that all edges in the graph point to existing vertices, and that there are
+// no dependency cycles. After validation, each vertex will be run, deterministically, in parallel
+// topological order. Difference between Run is it will continue execution despite vertex returning
+// an error (although will not execute it's dependencies).
+func (d *Runner) RunOnFailure() error {
+	// sanity check
+	if len(d.fns) == 0 {
+		return nil
+	}
+	// count how many deps each vertex has
+	deps := make(map[string]int)
+	for vertex, edges := range d.graph {
+		// every vertex along every edge must have an associated fn
+		if _, ok := d.fns[vertex]; !ok {
+			return errMissingVertex
+		}
+		for _, vertex := range edges {
+			if _, ok := d.fns[vertex]; !ok {
+				return errMissingVertex
+			}
+			deps[vertex]++
+		}
+	}
+
+	if d.detectCycles() {
+		return errCycleDetected
+	}
+
+	running := 0
+	resc := make(chan result, len(d.fns))
+	var err error
+
+	// start any vertex that has no deps
+	for name := range d.fns {
+		if deps[name] == 0 {
+			running++
+			start(name, d.fns[name], resc)
+		}
+	}
+
+	// wait for all running work to complete
+	for running > 0 {
+		res := <-resc
+		running--
+
+		// capture the first error
+		if res.err != nil && err == nil {
+			err = res.err
+		}
+		
+		if res.err == nil {
+			// start any vertex whose deps are fully resolved
+			for _, vertex := range d.graph[res.name] {
+				if deps[vertex]--; deps[vertex] == 0 {
+					running++
+					start(vertex, d.fns[vertex], resc)
+				}
+			}
+		}
```

Comment:
I don't think this check is in the right place.  Right now if a node terminates with an error, it won't check (at that time) whether to look for more nodes to run.  But if a node subsequently finishes successfully, a dependent of a failed node could still run.

Separately - where are things like `when: status: failure | success` handled?  Are you sure this go module is actually what's used by drone?

Reply:
Status filtering etc... is handled within the func that gets executed by the dagrunner see 
https://github.com/drone/runner-go/blob/17a63102bc1dd7ffbe6472f628ba0763228e85aa/pipeline/runtime/execer.go#L67

Reply:
Namely this function gets executed in the dag runner.
https://github.com/drone/runner-go/blob/17a63102bc1dd7ffbe6472f628ba0763228e85aa/pipeline/runtime/execer.go#L113

Reply:
@lily-commure I don't think I'm fulling following your scenario. Right now if it errs it won't execute any of it's dependents but other nodes running will still be looping through and running. It can't run any of it's dependents because their counter can never get decremented to zero.